File: Week002/acruikshank/goal.js
// Syntactic sugar for strings starting with 'Go*'
function G(s) {
  return (function o(p) {   // inner function curries prefix
    return function(s) {    // external function taking optional arg
      return s===undefined  // test arg
        ? o(p + 'o')        // generate function with longer prefix
        : p + s;            // return prefix plus string
  }})('G')( s )             // gen inner funct with 'G' prefix, call it with 1st arg
}

expect( G('') ).to.be( 'G' );
expect( G('al') ).to.be( 'Gal' );
expect( G()('al') ).to.be( 'Goal' );
expect( G()()()()()('al') ).to.be( 'Goooooal' );
expect( G()()('') ).to.be( 'Goo' );
expect( G()()('nies') ).to.be( 'Goonies' );

function expect(act) {
  return {to:{be:function(exp) {
    if (exp === act) console.log("PASS")
    else console.log("FAIL: Expected '"+act+"' to be '"+exp+"'");
  }}}
}

File: Week002/acruikshank/goal.hs
{-# LANGUAGE FlexibleInstances #-}

class GoalStart a where g :: a

instance GoalStart (String -> String) where
  g s = 'g' : s
instance GoalArg x => GoalStart (() -> x) where
  g u = goalarg "go"

class GoalArg a where goalarg :: String -> a

instance GoalArg x => GoalArg (() -> x) where
  goalarg a x = goalarg (a ++ "o")
instance GoalArg (String -> String) where
  goalarg a s = a ++ s

main = do
  putStrLn $ if "g" == g""                        then "PASS" else "FAIL"
  putStrLn $ if "gal" == g"al"                    then "PASS" else "FAIL"
  putStrLn $ if "goal" == g()"al"                 then "PASS" else "FAIL"
  putStrLn $ if "goooooal!" == g()()()()()"al!"   then "PASS" else "FAIL"
  putStrLn $ if "goo" == g()()""                  then "PASS" else "FAIL"
  putStrLn $ if "goonies" == g()()"nies"          then "PASS" else "FAIL"
